### PR TITLE
fix: new group guest link creation AR-3188

### PR DESCRIPTION
**Pull request by tmspzz**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/AR-3188" title="AR-3188" target="_blank">BugAR-3188</a>  When we create a group on AR, we can not enable guestlinks
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
    - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
    - [x] contains a reference JIRA issue number like `SQPIT-764`
    - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
    - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

1. When creating a new group, the UI wasn't using the correct accessRoles
1. When creating a new group and disallowing guests, the UI would have a race where some guests could remain in the list of participants and fail group creation
1. When turning off guest access, guest links are not revoked and can't be revoked when access is disabled

### Causes (Optional)

_Briefly describe the causes behind the issues. This could be helpful to understand the adopted solutions behind some nasty bugs or complex issues._

### Solutions

1. Remove redundant state information, so that there is only 1 way of determining accessRoles
1. Bulk remove guests and avoid triggering state updates. The problem arises from a misunderstanding of how the `collect` operator works
1. When disabling guests, also revoke guest link

#### How to Test

1. Create a a group with guests, check that you can create a guest link right after group creation
1. Start creating a group with guests, disallow guests before hitting continue and check that group creation worked
1. After creating a group with guests, edit guest access. Notice that the guest link is also revoked.

----
#### PR Post Submission Checklist for internal contributors (Optional)

- [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/details/editguestaccess/EditGuestAccessViewModel.kt [hunk 1]
```diff
@@ -162,6 +162,9 @@ class EditGuestAccessViewModel @Inject constructor(
 
         viewModelScope.launch {
             withContext(dispatcher.io()) {
+                if(!shouldEnableGuestAccess) {
```
Comment (tmspzz) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/details/editguestaccess/EditGuestAccessViewModel.kt]: When disallowing guests, revoke guest links as well

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/search/SearchPeopleViewModel.kt [hunk 2]
```diff
@@ -261,15 +261,31 @@ abstract class SearchPeopleViewModel(
...
+     * thread like to use outdated values.
+     * Use [removeContactsFromGroup] instead.
+     */
     fun removeContactFromGroup(contact: Contact) {
+        removeContactsFromGroup(setOf(contact))
+    }
+    fun removeContactsFromGroup(contacts: Set<Contact>) {
         viewModelScope.launch {
-            selectedContactsFlow.emit(selectedContactsFlow.value - contact)
+            selectedContactsFlow.emit(selectedContactsFlow.value - contacts)
```
Comment (tmspzz) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/search/SearchPeopleViewModel.kt:288]: @trOnk12

File: app/src/main/kotlin/com/wire/android/ui/home/newconversation/NewConversationViewModel.kt [hunk 3]
```diff
@@ -93,22 +93,10 @@ class NewConversationViewModel @Inject constructor(
 
     fun onAllowGuestStatusChanged(status: Boolean) {
         groupOptionsState = groupOptionsState.copy(isAllowGuestEnabled = status)
-        if (!status) {
```
Comment (tmspzz) [inline app/src/main/kotlin/com/wire/android/ui/home/newconversation/NewConversationViewModel.kt:96]: We don't need any `accessRoleState`. Simplify the logic here

File: app/src/main/kotlin/com/wire/android/ui/home/newconversation/groupOptions/GroupOptionState.kt [hunk 4]
```diff
@@ -30,10 +30,6 @@ data class GroupOptionState(
     val isAllowServicesEnabled: Boolean = true,
     val isReadReceiptEnabled: Boolean = true,
     val showAllowGuestsDialog: Boolean = false,
-    val accessRoleState: MutableSet<Conversation.AccessRole> = Conversation
```
Comment (tmspzz) [inline app/src/main/kotlin/com/wire/android/ui/home/newconversation/groupOptions/GroupOptionState.kt:33]: We don't need any `accessRoleState`. Simplify the logic here

**Comment by AndroidBob** [comment]
Build [2](https://10.10.124.134/job/wire-android-reloaded/job/PR-1558/2/) **failed**.

**Comment by AndroidBob** [comment]
Build [3](https://10.10.124.134/job/wire-android-reloaded/job/PR-1558/3/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv4.0.2(6006422)(6006422)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-vv4.0.2(6006422)(6006422)-staging-compat.apk)

**Comment by AndroidBob** [comment]
Build [4](https://10.10.124.134/job/wire-android-reloaded/job/PR-1558/4/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv4.0.2(6007766)(6007766)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-vv4.0.2(6007766)(6007766)-staging-compat.apk)

**Review by alexandreferris — APPROVED** [review]
Looking great!

Tested out on Pixel 4 API 33 and its working as expected.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/details/editguestaccess/EditGuestAccessViewModel.kt [hunk 5]
```diff
@@ -162,6 +162,9 @@ class EditGuestAccessViewModel @Inject constructor(
 
         viewModelScope.launch {
             withContext(dispatcher.io()) {
+                if (!shouldEnableGuestAccess) {
+                    removeGuestLink()
```
Comment (gongracr) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/details/editguestaccess/EditGuestAccessViewModel.kt:166]: Then switching the guests allowed switch, will trigger the display of the revoke guest link?? Would that be ok from product perspective or a bit confusing?

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/search/SearchPeopleViewModel.kt [hunk 6]
```diff
@@ -261,15 +261,31 @@ abstract class SearchPeopleViewModel(
         }
     }
 
+    /**
+     * Do not under any circumstances call this function in a loop.
+     * It will cause the to emit a value each time. This in turn
+     * will result in synchronous execution of functions on the main
+     * thread like to use outdated values.
+     */
```
Comment (gongracr) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/search/SearchPeopleViewModel.kt:269]: Then we should also rename it accordingly to increase redundancy to something like `addSingleContactToGroup(contact: Contact)` ?

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/search/SearchPeopleViewModel.kt [hunk 7]
```diff
@@ -261,15 +261,31 @@ abstract class SearchPeopleViewModel(
...
         }
     }
 
+    /**
+     * Do not under any circumstances call this function in a loop.
+     * It will cause the state to emit a value each time. This in turn
+     * will result in synchronous execution of functions on the main
+     * thread like to use outdated values.
+     * Use [removeContactsFromGroup] instead.
+     */
```
Comment (gongracr) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/search/SearchPeopleViewModel.kt:282]: Then we should also rename it accordingly for redundancy to something like removeSingleContactFromGroup(contact: Contact) ?

**Review by gongracr — APPROVED** [review]
Just some ux and renaming comments. But looking much clearer now. GJ 🍎

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/search/SearchPeopleViewModel.kt (hunk 6 above)
Comment (tmspzz) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/search/SearchPeopleViewModel.kt:269]: It is singular. ContacT

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/search/SearchPeopleViewModel.kt (hunk 7 above)
Comment (tmspzz) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/search/SearchPeopleViewModel.kt:282]: It is singular. ContacT

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/search/SearchPeopleViewModel.kt (hunk 6 above)
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/search/SearchPeopleViewModel.kt:269]: Exactly, so with the `xxxSinglexxx` modifier, will be even more clear that this function shouldn't be called from a loop imo.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/search/SearchPeopleViewModel.kt (hunk 7 above)
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/search/SearchPeopleViewModel.kt:282]: same than above

**Pull request merged**